Add abstract hardware interface and refactor Robot Driver

jcnorby opened the pull request:
Changes include:
- Add HardwareInterface abstract class for extensibility
- Switch MBLinkConverter to SpiritInterface which inherits HardwareInterface
- Refactor robot_driver package to include directories for hardware interfaces and controllers
- Remove leg override node and support (deprecated)
- Remove open loop controller package (deprecated)
- Remove robot_proxy and ground_truth_publisher (deprecated)

These changes have been tested in sim but not yet in hardware. I would recommend we merge this PR sooner rather than later to permit additional refactoring, then test in hardware at a later date and submit PRs/hotfixes to resolve any remaining bugs.

jcnorby commented on robot_driver/src/controllers/leg_controller.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "robot_driver/controllers/leg_controller.h"
```
Note this is not a new file, just renamed from leg_controller_template.cpp

justinyim commented on robot_driver/include/robot_driver/hardware_interfaces/hardware_interface.h:
```diff
@@ -0,0 +1,68 @@
+   * @param[out] joint_state_msg Message containing joint state information
+   * @param[out] imu_msg Message containing imu information
+   */
+  virtual bool recv(sensor_msgs::JointState& joint_state_msg,
```
Should we just let users add additional variables if they want to use additional data from their robots (A1 foot contact sensors, other platforms' cameras or odometry handled by the built-in computers)?

jcnorby replied:
Added Eigen::VectorXd user_rx_data (in addition to user_tx_data).

justinyim commented on robot_driver/src/robot_driver.cpp:
```diff
@@ -64,18 +65,19 @@ RobotDriver::RobotDriver(ros::NodeHandle nh, int argc, char** argv) {
   // Set up pubs and subs dependent on robot layer
-  if (is_hw_) {
+  if (is_hardware_) {
     ROS_INFO("Loading hw robot driver");
```
Expand hw to hardware?

justinyim reviewed (approved):
Looks good to me

jcnorby commented:
Just tested and confirmed that this doesn't work in hardware yet, something causes mblink_.start(argc,argv) to segfault. Do we want to approve and fix it later or try to iron that out first?

ardalantj commented:
Lets try and fix it if its segfaulting. I worry we wouldnt get to it later if its merged now.

jcnorby commented:
These changes have now been tested in hardware, they require a clean catkin install so that catkin can handle the new paths correctly. I also updated the topic names for the imu and joint data, added a publisher for the joint data, and removed ground_truth_publisher and robot_proxy as they are no longer needed.
